### ISSUE TITLE
-iu flag slows down closest command significantly

### ISSUE THREAD

**Issue by jma1991** [arq5x/bedtools2#268]
I'm trying to identify the closest regions in my target file which are downstream, on the same strand, and do not overlap those in my query file. In addition, I want to report the distance. I'm using the following command:

``` bash
bedtools closest -a reads.bed -b regions.bed -s -io -iu -D a > reads.closest.regions.bed
```

Both the reads.bed and regions.bed files have ~20 million records. When I run this command it runs without error, but has still not finished ~ 2 hours later. When I run variations of this command, they only take ~5-10 minutes:

``` bash
time bedtools closest -a reads.bed -b regions.bed -s > /dev/null
real    9m5.165s
user    8m57.597s
sys 0m6.920s
```

``` bash
time bedtools closest -a reads.bed -b regions.bed -s -io > /dev/null
real    5m25.619s
user    5m21.876s
sys 0m3.191s
```

``` bash
time bedtools closest -a reads.bed -b regions.bed -s -io -D a > /dev/null
real    5m40.379s
user    5m36.784s
sys 0m3.019s
```

I've tried to make sure my BED file is correct by using bedops error checking:

``` bash
bedops --ec --everything reads.bed 1> /dev/null
bedops --ec --everything regions.bed 1> /dev/null
```

Does ignoring records upstream of the query record via the -iu flag introduce this much slow down or is this a possible bug? I'm using bedtools v2.24.0.

**Comment by nkindlon** [comment]
Hi James,
    The good news is that, since you definitely want the downstream hits, you can use the -fd option ("Force Downstream"), which will consider the downstream hits first. I'm fairly sure that should get you better performance.

     The bad news is that I'm no longer with the lab, and Aaron's dealing with a lot of other stuff right now, so I don't know when this can be properly debugged and fixed. But below are some suggestions to whoever looks at it (me, him, or someone else):

In src/utile/NewChromsweep/CloseSweep.cpp:

Change line 404 from:
if (totalHitsUsed < _kClosest && !overlaps->empty()) {
to 
if (!_context->ignoreOverlaps() && totalHitsUsed < _kClosest && !overlaps->empty()) {

Similarly, change line 430 from:
if (upDist < downDist || (tie && tieMode != ContextClosest::LAST_TIE)) {
to
if (!_context->ignoreUpstream() && (upDist < downDist || (tie && tieMode != ContextClosest::LAST_TIE))) {

and likewise line 435 from:
if (downDist < upDist || (tie && tieMode != ContextClosest::FIRST_TIE)) {
to 
if (!_context->ignoreDownstream() && (downDist < upDist || (tie && tieMode != ContextClosest::FIRST_TIE))) {

**Comment by arq5x** [comment]
Hi James,

Once we get a breather from the current onslaught (likely mid August), we will dig into this to find the source of the slowdown.  Thanks very much for reporting this.

**Comment by nkindlon** [comment]
Looking at this now. I'm able to reproduce the error in the latest repository with 2 files each of 100K reads only on chr1 (generated by the bedtools random tool). I suspect even smaller files will show it as well.

$ ~/bedtools/bugs/bug268$ time ~/bedtools/workspace/bt2-release/bin/bedtools closest -a a100K.txt -b b100K.txt -s -io -D a > /dev/null

real	0m0.192s
user	0m0.176s
sys	0m0.016s

$ ~/bedtools/bugs/bug268$ time ~/bedtools/workspace/bt2-release/bin/bedtools closest -a a100K.txt -b b100K.txt -s -io -D a -iu > /dev/null

(doesn't finish after several minutes.)

**Comment by nkindlon** [comment]
This is "mostly fixed" in pull request #294. What I mean is that I established that -iu was running very slow only when used with -s and -S. This is now fixed. 
But there was a similar undiscovered problem that -id would also run slow if used with -s or -S. I fixed that also, but only with -D ref. If using -D a or -D b with -id and -s or -S, it will still be slow. Aaron, I'll shoot you a mail with more details.
